src/hnsw/product_quantization.c

```diff
@@ -0,0 +1,275 @@
+            if(distance < min_dist) {
+                min_dist = distance;
+                min_dist_cluster_idx = j;
+            }
+        }
+        current_cluster = clusters[ min_dist_cluster_idx ];
+
+        if(!current_cluster->point_cnt) {
+            // This is not optimal but as long as this is just array of pointers
+            // It won't consume much memory
+            current_cluster->points = palloc(sizeof(size_t) * dataset_size);
+        }
+        current_cluster->points[ current_cluster->point_cnt ] = current_subset;
+        current_cluster->point_cnt += 1;
+    }
+}
+
+/*
+ * Calculate mean value of given 2d array
+ * */
+static float4 *calculate_mean(float4 **points, uint32 point_cnt, uint32 subset_dim)
+{
+    float4 *mean = palloc0(sizeof(float4) * subset_dim);
+    uint32  i;  // rows
+    uint32  j;  // cols
+
+    for(i = 0; i < point_cnt; i++) {
+        for(j = 0; j < subset_dim; j++) {
+            if(!mean[ j ]) mean[ j ] = 0;
+            mean[ j ] += points[ i ][ j ];
+        }
+    }
+
+    for(i = 0; i < subset_dim; i++) {
+        mean[ i ] = mean[ i ] / point_cnt;
+    }
+
+    return mean;
+}
+
+/*
+ * Update cluster centers with the mean values of assigned points
+ * And remove points from cluster
+ * */
+static void update_centers(Cluster **clusters, uint32 k, uint32 subset_dim)
+{
+    uint32   i;
+    Cluster *current_cluster;
+
+    for(i = 0; i < k; i++) {
+        current_cluster = clusters[ i ];
+        if(!current_cluster->point_cnt) continue;
+
+        current_cluster->center = calculate_mean(current_cluster->points, current_cluster->point_cnt, subset_dim);
+
+        current_cluster->points = NULL;
+        current_cluster->point_cnt = 0;
+    }
+}
+
+/*
+ * Checks whether distances between new and old centers
+ * are less then the defined threshold to stop iterations
+ * and return current centers as result.
+ * */
+bool should_stop_iterations(float4              **old_centers,
+                            Cluster             **clusters,
+                            uint32                cluster_count,
+                            uint32                subset_dim,
+                            usearch_metric_kind_t distance_metric)
+{
+    usearch_error_t error = NULL;
+    uint32          i;
+    float4          threshold = 0.15f;
+    float4          distance = 0.0f;
+
+    for(i = 0; i < cluster_count; i++) {
+        distance += usearch_distance(
+            old_centers[ i ], clusters[ i ]->center, usearch_scalar_f32_k, subset_dim, distance_metric, &error);
+        assert(!error);
+    }
+
+    distance /= cluster_count;
+
+    return distance <= threshold;
+}
+
+/*
+ * Run k-means clustering over the dataset in a subset of vectors
+ * args:
+ *   k - cluster count
+ *   dataset - 2d array of float numbers
+ *   dataset_size - length of dataset (axis 0)
+ *   subset_start - index from which the subset will be taken (in case if subset count is 1 this will be 0)
```

Comment:
> in case if subset count is 1 this will be 0

what does this mean?

Reply:
If let's say we are quantizing [f32x128] -> [u8x1] so the only range will be [0-128] (subvector_start: 0, subvector_dim: 128), but I will rephrase this comment as for the first subvector it will also be 0